packages/core/src/awsService/cloudWatchLogs/commands/tailLogGroup.ts

```diff
@@ -3,17 +3,142 @@
+}
+
+export function closeSession(sessionUri: vscode.Uri, registry: LiveTailSessionRegistry) {
+    const session = registry.get(sessionUri)
+    if (session === undefined) {
+        throw new ToolkitError(`No LiveTail session found for URI: ${sessionUri.toString()}`)
+    }
+    session.stopLiveTailSession()
+    registry.delete(sessionUri)
+}
+
+export async function clearDocument(textDocument: vscode.TextDocument) {
+    const edit = new vscode.WorkspaceEdit()
+    const startPosition = new vscode.Position(0, 0)
+    const endPosition = new vscode.Position(textDocument.lineCount, 0)
+    edit.delete(textDocument.uri, new vscode.Range(startPosition, endPosition))
+    await vscode.workspace.applyEdit(edit)
+}
+
+async function prepareDocument(session: LiveTailSession): Promise<vscode.TextDocument> {
+    const textDocument = await vscode.workspace.openTextDocument(session.uri)
+    await clearDocument(textDocument)
+    await vscode.window.showTextDocument(textDocument, { preview: false })
+    await vscode.languages.setTextDocumentLanguage(textDocument, 'log')
+    return textDocument
+}
+
+async function handleSessionStream(
+    stream: AsyncIterable<StartLiveTailResponseStream>,
+    document: vscode.TextDocument,
+    session: LiveTailSession
+) {
+    try {
+        for await (const event of stream) {
+            if (event.sessionUpdate !== undefined) {
+                const formattedLogEvents = event.sessionUpdate.sessionResults!.map<string>((logEvent) =>
+                    formatLogEvent(logEvent)
+                )
+                if (formattedLogEvents.length !== 0) {
+                    await updateTextDocumentWithNewLogEvents(formattedLogEvents, document, session.maxLines)
+                }
+            }
+        }
+    } catch (err) {
+        throw new ToolkitError('Caught on-stream exception')
+    }
+}
+
+function formatLogEvent(logEvent: LiveTailSessionLogEvent): string {
+    if (!logEvent.timestamp || !logEvent.message) {
+        return ''
+    }
+    const timestamp = new Date(logEvent.timestamp).toLocaleTimeString('en', {
+        timeStyle: 'medium',
+        hour12: false,
+        timeZone: 'UTC',
+    })
+    let line = timestamp.concat('\t', logEvent.message)
+    if (!line.endsWith('\n')) {
+        line = line.concat('\n')
+    }
+    return line
+}
+
+async function updateTextDocumentWithNewLogEvents(
+    formattedLogEvents: string[],
+    document: vscode.TextDocument,
+    maxLines: number
+) {
+    const edit = new vscode.WorkspaceEdit()
+    formattedLogEvents.forEach((formattedLogEvent) =>
+        edit.insert(document.uri, new vscode.Position(document.lineCount, 0), formattedLogEvent)
+    )
+    await vscode.workspace.applyEdit(edit)
+}
+
+function registerTabChangeCallback(
+    session: LiveTailSession,
+    registry: LiveTailSessionRegistry,
+    document: vscode.TextDocument
+) {
+    vscode.window.tabGroups.onDidChangeTabs((tabEvent) => {
```

Comment:
If i'm understanding this correctly I think you might want vscode.workspace.textDocuments instead. I've never actually seen vscode.window.tabGroups before :thinking:. The process would be more or less the same expect instead of onDidChangeTabs and interating over every tab group you can just. Something like:

```
vscode.workspace.onDidChangeTextDocument((textDocument) => {
    if (textDocument.document.uri.toString() === liveTailSession.uri.toString()) {

    }
})
```

Reply:
What we are trying to achieve is that if a user doesn't have the LiveTail session open in any tab in their editor, that we close the session.

What I've learned is that the underlying TextDocument can remain open, even if its editor tabs are closed. It seems that the LiveTail command will keep the doucment open even if all its editors are closed because the command is still writing from the stream to the document. https://code.visualstudio.com/api/references/vscode-api#workspace.onDidCloseTextDocument. This suggests looking at onDidChangeVisibleTextEditors, but that event fires even when swapping tabs, not closing. https://code.visualstudio.com/api/references/vscode-api#window.onDidChangeVisibleTextEditors

This implementation is admittedly heavy, but has been the only thing I've found that fits what I'm trying to accomplish.